src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,275 @@
+            let headers_len = self
+                .headers_iter
+                .clone()
+                .fold(0, |acc, (_k, v)| acc + v.len());
+            let header = IndexHeader::new(record_header_size, headers_len, meta.len());
+            let key_size = key.as_ref().len();
+            Ok(HeaderStage {
+                headers_iter: self.headers_iter,
+                header,
+                key_size,
+                meta,
+            })
+        } else {
+            Err(anyhow!("BTree is empty, can't find info about key len!"))
+        }
+    }
+}
+
+impl<'a, K: AsRef<[u8]>, V: Serialize + 'a, It: Iterator<Item = (K, &'a Vec<V>)> + Clone>
+    HeaderStage<'a, K, V, It>
+{
+    pub(super) fn leaves_stage(self) -> Result<LeavesStage<'a, K, V, It>> {
+        let hs = self.header.serialized_size()? as usize;
+        let fsize = self.header.meta_size;
+        let msize: usize = TreeMeta::serialized_size_default()? as usize;
+        let headers_start_offset = (hs + fsize) as u64;
+        let headers_size = self.header.records_count * self.header.record_header_size;
+        let leaves_offset = headers_start_offset + (headers_size + msize) as u64;
+        // leaf contains pair: (key, offset in file for first record's header with this key)
+        let leaf_size = self.key_size + std::mem::size_of::<u64>();
+        let leaves_buf = Self::serialize_leaves(
+            self.headers_iter.clone(),
+            headers_start_offset,
+            leaf_size,
+            self.header.record_header_size,
+        );
+        Ok(LeavesStage {
+            headers_iter: self.headers_iter,
+            leaf_size,
+            leaves_buf,
+            leaves_offset,
+            headers_size,
+            meta: self.meta,
+            header: self.header,
+        })
+    }
+
+    fn serialize_leaves(
+        headers_iter: It,
+        headers_start_offset: u64,
+        leaf_size: usize,
+        record_header_size: usize,
+    ) -> Vec<u8> {
+        let mut leaves_buf = Vec::with_capacity(leaf_size * headers_iter.clone().count());
+        headers_iter.fold(
+            (headers_start_offset, &mut leaves_buf),
+            |(offset, leaves_buf), (leaf, hds)| {
+                leaves_buf.extend_from_slice(leaf.as_ref());
+                let offsetb = serialize(&offset).expect("serialize u64");
+                leaves_buf.extend_from_slice(&offsetb);
+                let res = (offset + (hds.len() * record_header_size) as u64, leaves_buf);
+                res
+            },
+        );
+        leaves_buf
+    }
+}
+
+impl<'a, K: AsRef<[u8]>, V: Serialize + 'a, It: Iterator<Item = (K, &'a Vec<V>)> + Clone>
+    LeavesStage<'a, K, V, It>
+{
+    pub(super) fn tree_stage(self) -> Result<TreeStage<'a, K, V, It>> {
+        // FIXME: remove collect (now it is used to iterate through `chunks` in `serialize_bptree`)
+        let keys: Vec<K> = self.headers_iter.clone().map(|e| e.0).collect();
```

Comment:
Agree, `collect()` is redundant, you can pass `impl Iterator` into `serialize_bptree()`

Reply:
now keys are not collected